File: apps/android/app/src/main/java/com/exponential/app/domain/AgentComposerPrompt.kt
package com.exponential.app.domain

/**
 * EXP-825: the composer's free text on the wire. With a subject picked it is
 * OPTIONAL additional instructions (the desktop renders it as an "Additional
 * instructions from the requester" section); with none it IS the chat prompt,
 * and for the Chat and Create action builtins the server requires it. Images
 * ride it as the steer embed format ([buildSteerImageMessage], byte-identical
 * ×4): prose, blank line, one `![image](/api/attachments/<id>)` line per
 * upload, `[Image #k]` markers in the prose. The contract caps both
 * (`startPrompt`). Mirrors iOS `AgentComposerPrompt`.
 */
object AgentComposerPrompt {

    /** The contract's `startPrompt.maxLength` (16384). */
    const val MAX_LENGTH: Int = DomainContract.startPromptMaxLength

    /** The contract's `startPrompt.maxImages` — the same four the steer
     * composer carries ([MAX_STEER_IMAGES]). */
    const val MAX_IMAGES: Int = DomainContract.startPromptMaxImages

    /**
     * The `prompt` to send: null when there is nothing to say (blank text and
     * no images), so the wire omits the key and the server sees NO prompt
     * rather than an empty one.
     */
    fun build(text: String, attachmentIds: List<String>): String? {
        val trimmed = text.trim()
        if (trimmed.isEmpty() && attachmentIds.isEmpty()) return null
        return buildSteerImageMessage(trimmed, attachmentIds)
    }

    /**
     * Whether the COMPOSED prompt — the trimmed draft plus one embed line per
     * pending image — fits the contract cap; the server measures the whole
     * string (`startPromptSchema`), not the prose. Attachment ids are UUIDs,
     * so a fixed-width stand-in measures the embeds exactly before the
     * upload has minted them.
     */
    fun withinLimit(text: String, imageCount: Int = 0): Boolean =
        (build(text, List(imageCount) { ATTACHMENT_ID_STAND_IN })?.length ?: 0) <= MAX_LENGTH

    /** A 36-char UUID-shaped placeholder for a not-yet-uploaded attachment. */
    private const val ATTACHMENT_ID_STAND_IN = "00000000-0000-0000-0000-000000000000"

    /** What the composer is about to start. */
    sealed interface Subject {
        /** No subject: a chat. */
        data object None : Subject

        /** [count] checked issues (1 = a single-issue run, 2+ = a batch). */
        data class Issues(val count: Int) : Subject

        /** A team action or a builtin. */
        data object Action : Subject
    }

    /**
     * The submit label — the ×4 contract: "Start chat" / "Start coding" /
     * "Start batch · N" / "Run action".
     */
    fun submitTitle(subject: Subject): String = when (subject) {
        Subject.None -> "Start chat"
        is Subject.Issues -> if (subject.count > 1) "Start batch · ${subject.count}" else "Start coding"
        Subject.Action -> "Run action"
    }
}

File: apps/ios/ExpCore/Sources/Domain/AgentComposerPrompt.swift
import Foundation

/// EXP-825: the composer's free text on the wire. With a subject picked it is
/// OPTIONAL additional instructions (the desktop renders it as an
/// "Additional instructions from the requester" section); with none it IS
/// the chat prompt, and for the Chat and Create action builtins the server
/// requires it. Images ride it as the steer embed format
/// (`SteerImageMessage.build`, byte-identical ×4): prose, blank line, one
/// `![image](/api/attachments/<id>)` line per upload, `[Image #k]` markers in
/// the prose. The contract caps both (`startPrompt`).
public enum AgentComposerPrompt {
    /// The contract's `startPrompt.maxLength` (16384, UTF-16 code units on
    /// the web — `count` here counts characters, which can only be FEWER).
    public static let maxLength = DomainContract.startPromptMaxLength
    /// The contract's `startPrompt.maxImages` — the same four the steer
    /// composer carries.
    public static let maxImages = DomainContract.startPromptMaxImages

    /// The `prompt` to send: nil when there is nothing to say (blank text and
    /// no images), so the wire omits the key and the server sees NO prompt
    /// rather than an empty one.
    public static func build(text: String, attachmentIds: [String]) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty, attachmentIds.isEmpty { return nil }
        return SteerImageMessage.build(text: trimmed, attachmentIds: attachmentIds)
    }

    /// Whether the draft (before its embeds) fits the contract cap.
    public static func withinLimit(_ text: String) -> Bool {
        text.count <= maxLength
    }

    /// What the composer is about to start.
    public enum Subject: Equatable, Sendable {
        /// No subject: a chat.
        case none
        /// `count` checked issues (1 = a single-issue run, 2+ = a batch).
        case issues(count: Int)
        /// A team action or a builtin.
        case action
    }

    /// The submit label — the ×4 contract: "Start chat" / "Start coding" /
    /// "Start batch · N" / "Run action".
    public static func submitTitle(for subject: Subject) -> String {
        switch subject {
        case .none: "Start chat"
        case let .issues(count): count > 1 ? "Start batch · \(count)" : "Start coding"
        case .action: "Run action"
        }
    }

    /// The field's placeholder — web `composerPlaceholder` byte for byte
    /// (EXP-825): a chat asks for the message; a picked action with a
    /// non-blank `promptPlaceholder` (`actionHint`, trimmed) shows it; every
    /// other subject asks for what is optional next to it. Issue chips never
    /// read an action's hint.
    public static func placeholder(for subject: Subject, actionHint: String?) -> String {
        switch subject {
        case .none:
            return "Ask the agent…"
        case .action:
            let hint = actionHint?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return hint.isEmpty ? "Additional instructions (optional)…" : hint
        case .issues:
            return "Additional instructions (optional)…"
        }
    }
}
